### PR TITLE
Support 2023.3 IDEA Version and Fix Minor Error False Positives

### LINKED ISSUE #151: `:=` not allowed after `else` keyword
# Actual Behavior

image

Plugin reports an error for `:=` assignment after `else` keyword.

# Expected  Behavior
No error reported as shown here: https://play.openpolicyagent.org/p/AfCBX28GQh

# How To Reproduce
```
package play

test := "allow" {
    false
} else := "deny"
```

# Environment
 - OS: MacOS
 - OPA version: v0.40.0 or newer
 - IDE full name and version: IntelliJ IDEA 2023.1.2 (Ultimate Edition) Build #IU-231.9011.34, built on May 16, 2023
 - plugin version: v0.10.0

# Additional context
`opa-idea-plugin` uses grammar for OPA from version older than v0.40.0. In version v0.40.0 the grammar was changed to allow `:=` assignment operator after `else` keyword: https://www.openpolicyagent.org/docs/v0.40.0/policy-reference/#grammar

# Suggested labels
#bug #area/grammar

### LINKED ISSUE #165: Not compatible with WebStorm 2023.3.3
Not sure to classify this as a bug or a FR.

# Actual Behavior

- Plugin is not compatible with 2023.3.2 version of JetBrains Intellij

# Expected  Behavior

- Plugin is compatible with 2023.3.2 version of JetBrains Intellij

# How To Reproduce
1. Update your Intellij IDE to latest
2. Go to https://plugins.jetbrains.com/plugin/14865-open-policy-agent
3. See error
Screenshot 2024-02-01 at 12 44 52 pm

Alternatively upgrade your Webstorm and see the below error in the bottom right hand corner:

```
Plugin 'Open Policy Agent' (version 'v0.11.0') is not compatible with the current version of the IDE, because it requires build 232.* or older but the current build is WS-233.14015.89
```

# Environment
 - OS: N/A

 - OPA version: N/A

 - IDE full name and version:
 WebStorm 2023.3.3
Build #WS-233.14015.89, built on January 23, 2024

 - plugin version: v0.11.0

### PR DESCRIPTION
**Pull request by spoage**
# Description
* Compatibility: upgrade to 2023.3 / 233.2 version
* Grammar: rules with multidimensional array heads support
* Grammar: else expressions no longer consider ':=' invalid
* tests: bump OPA version to v0.61.0

 

Fixes #165
Fixes #151

### CODE REVIEW

File: src/main/grammar/Rego.bnf [hunk 1]
```diff
@@ -67,10 +67,10 @@ module              ::= package (import|  rule)*
 package             ::= "package" ref
 import              ::= "import" ref ( "as" var )?
 rule                ::= "default"?  rule-head  rule-body*
-rule-head           ::= var ( "(" rule-args? ")" )? ("[" term "]" )? ( ( ":=" | "=" ) expr )?
+rule-head           ::= var ( "(" rule-args? ")" )? ("[" term "]" )* ( ( ":=" | "=" ) expr )?
```
Comment (anderseknert) [inline src/main/grammar/Rego.bnf:70]: What does this change mean in practice?

File: src/main/grammar/Rego.bnf [hunk 2]
```diff
@@ -67,10 +67,10 @@ module              ::= package (import|  rule)*
...
 rule-args           ::= term ( "," term )*
 rule-body           ::= else-expr  | query-block
-else-expr           ::= else "=" expr  query-block? | else "="? query-block
+else-expr           ::= else (':=' | '=') expr  query-block? | else "="? query-block
```
Comment (anderseknert) [inline src/main/grammar/Rego.bnf:73]: I think this should be updated in the last expression too? i.e. the `else "="?`

File: src/test/resources/org/openpolicyagent/ideaplugin/lang/parser/fixtures/else_keyword.rego [hunk 3]
```diff
@@ -56,6 +56,12 @@ check_only_equal_querry_for_else = true {
     true
 }
 
+check_else_with_colon_assignment_op {
+    input.x < input.y
+} else := {
+    true
+}
```
Comment (anderseknert) [inline src/test/resources/org/openpolicyagent/ideaplugin/lang/parser/fixtures/else_keyword.rego:63]: This is assigning a new set containing `true` as its only item. `else := true` would probably make more sense here.

**Review by anderseknert — COMMENTED** [review]
Thanks! Just some nits / questions.

File: src/main/grammar/Rego.bnf (hunk 1 above)
Comment (spoage) [reply src/main/grammar/Rego.bnf:70]: It allows multidimensional arrays for the rule head, as an alternative to the dot notation. See the added test in the `rule_head` test file for an example.

File: src/main/grammar/Rego.bnf (hunk 2 above)
Comment (spoage) [reply src/main/grammar/Rego.bnf:73]: Perhaps. I couldn't find a test case that I could test it on, so I chose to leave it alone. I can tweak it anyway, if desired?

Comment (spoage) [reply src/main/grammar/Rego.bnf]: I have replicated the addition to the other expression and added a test to cover it.

File: src/main/grammar/Rego.bnf (hunk 1 above)
Comment (anderseknert) [reply src/main/grammar/Rego.bnf:70]: Ah, gotcha! Thanks for bearing with me :)

File: src/main/grammar/Rego.bnf (hunk 2 above)
Comment (anderseknert) [reply src/main/grammar/Rego.bnf]: Thanks

**Review by anderseknert — APPROVED** [review]

**Comment by anderseknert** [comment]
Thanks Shane! Much appreciated. Now just to try and figure out how to get this published, as I've never been involved in that before 😅 If you have experience publishing a plugin and could help provide some pointers, hit me up on the OPA Slack!

**Comment by vgramer** [comment]
@spoage Thanks for the work :) 

Just a  clarification,  this PR fixes compatibility for  IDEA 2024.1 ( aka 241 in gradle, the new EAP) version on dev was already compatible with version 2023.3 (aka 233 in gradle).


@anderseknert, I have a working PR that fixes the release-note generation and publishing of the plugin... But it is not opened yet as I have weird behavior.  
TLDR:  [the last version on dev](https://plugins.jetbrains.com/plugin/14865-open-policy-agent/versions/dev/470424) should not have been published as tests failed due to missing opa binary. 

I will open PR this afternoon or tomorrow.

**Comment by anderseknert** [comment]
Ah, very happy to hear that Vincent! 👍 Looking forward to having Rego in my IDEA IDE's once more :)
